### PR TITLE
[FLINK-5265] Introduce state handle replication mode for CheckpointCoordinator

### PR DESCRIPTION
**Pull request by StefanRRichter**
Currently, the ``CheckpointCoordinator`` only supports repartitioning of ``OperatorStateHandle``s based on a split-and-distribute strategy. For future state types, such as broadcast or union state, we need a different repartitioning method that allows for replicating state handles to all subtasks.

This functionality is introduced with this PR and represents the first step on the way to implementing broadcast and union states.

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @aljoscha @StephanEwen

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/OperatorStateCheckpointOutputStream.java [hunk 1]
```diff
@@ -66,8 +66,10 @@ OperatorStateHandle closeAndGetHandle() throws IOException {
 			startNewPartition();
 		}
 
-		Map<String, long[]> offsetsMap = new HashMap<>(1);
-		offsetsMap.put(DefaultOperatorStateBackend.DEFAULT_OPERATOR_STATE_NAME, partitionOffsets.toArray());
+		Map<String, OperatorStateHandle.StateMetaInfo> offsetsMap = new HashMap<>(1);
+		OperatorStateHandle.StateMetaInfo metaInfo =
+				new OperatorStateHandle.StateMetaInfo(partitionOffsets.toArray(), OperatorStateHandle.Mode.BROADCAST);
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/OperatorStateCheckpointOutputStream.java]: I think this should be `SPLIT_REDISTRIBUTE` to conform to the old behaviour.

Comment (StefanRRichter) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/OperatorStateCheckpointOutputStream.java]: Yes, that is correct :-). Will change this as suggested.

**Comment by aljoscha** [comment]
Thanks for your work! 👍 I merged, could you please close this PR?

**Pull request closed without merging**